radicle-cli/tests/commands.rs

```diff
@@ -341,6 +341,12 @@ fn test_replication_via_seed() {
     alice.connect(&seed);
     bob.connect(&seed);
 
+    // Track node to enable fetch.
+    alice
+        .handle
+        .track_node(seed.id, Some("seed".to_string()))
+        .unwrap();
+
```

Comment:
Although this test is passing, I'm not sure why "seed" seems to be the announcer for bob's refs. Is that expected?

Reply:
It's because of the topology of the network:
```
alice -- seed -- bob
```

So when bob announces changes the seed hears them and relays them to alice. Although, I don't think having to track the seed makes much sense. It should be that alice tracks bob and hears about bob's refs so she'll replicate 🤔

Reply:
Yeah this is a good point. I'm guessing what's happening is that the seed is fetching from Alice, then the seed is making its own announcement of refs after it fetched, and Bob is fetching from the seed because he tracks it.

Probably we should be relaying announcements even if we don't follow the peer :thinking:.

Reply:
Ya, I would think that gossip should definitely be relayed. Otherwise we might end up in a scenario where a limited set of resources are tracked and announced :)

Reply:
My question is about the last leg of the replication, where Bob has created (cloned) its own refs under their namespace. Then Alice would replicate Bob's refs via the seed.  In this operation, Alice gets the refs announcement from the seed, which is expected. However, the announcement's "announcer" is set to the seed, not Bob, which I thought was the origin. 

Here is test log from Alice's point of view, where Alice's public key is:  `z6MknSLrJoTcukLrE435hVNQT4JUhbvWLX4kUzqkEStBU8Vi `,   the seed is `z6Mkux1aUQD2voWWukVb5nNUR7thrHveQG4pDQua8nVhib7Z`,  and Bob is `z6Mkt67GdsW7715MEfRuP4pSZxJRJh6kj6Y48WRqVv4N1tRk`: 

```
z6MknSLrJoTcukLrE435hVNQT4JUhbvWLX4kUzqkEStBU8Vi     service: Received message Announcement(z6Mkux1aUQD2voWWukVb5nNUR7thrHveQG4pDQua8nVhib7Z, Refs(rad:z42hL2jL4XNk6K8oHQaSWfMgCL7ji, [(PublicKey(z6Mkt67GdsW7715MEfRuP4pSZxJRJh6kj6Y48WRqVv4N1tRk), SignedRefs { refs: Refs({RefString("refs/heads/master"): Oid(f2de534b5e81d7c6e2dcaf58c3dd91573c0a0354), RefString("refs/rad/id"): Oid(3f3596fd2ce9a1e5c32b29840c1c2d38d85cd134)}), signature: Signature(z9So7T4ji6kibR1XyPvX97k86EFGYMmcZ3HEhe3RPSx2NpdrpMsrjiCc1vg3oLEswqqDrSMi6dRWQRQKfhBhaeDx), _verified: PhantomData<radicle_crypto::Unverified> }), (PublicKey(z6MknSLrJoTcukLrE435hVNQT4JUhbvWLX4kUzqkEStBU8Vi), SignedRefs { refs: Refs({RefString("refs/heads/master"): Oid(f2de534b5e81d7c6e2dcaf58c3dd91573c0a0354), RefString("refs/rad/id"): Oid(3f3596fd2ce9a1e5c32b29840c1c2d38d85cd134)}), signature: Signature(z2HaXR5GjG2UBXNjLBeCNSb6aicbGdiRaASm6ocLziDb7WfpRgQTgCruFyGSeRLoFiRc6H6VT1szb7M3b17WdRbez), _verified: PhantomData<radicle_crypto::Unverified> })])) from z6Mkux1aUQD2voWWukVb5nNUR7thrHveQG4pDQua8nVhib7Z          
``` 

later,  we can see Alice stored the refs from this announcement:

```
z6MknSLrJoTcukLrE435hVNQT4JUhbvWLX4kUzqkEStBU8Vi      worker: Git:  * [new ref]         refs/namespaces/z6Mkt67GdsW7715MEfRuP4pSZxJRJh6kj6Y48WRqVv4N1tRk/refs/heads/master -> refs/namespaces/z6Mkt67GdsW7715MEfRuP4pSZxJRJh6kj6Y48WRqVv4N1tRk/refs/heads/master
z6MknSLrJoTcukLrE435hVNQT4JUhbvWLX4kUzqkEStBU8Vi      worker: Git:  * [new ref]         refs/namespaces/z6Mkt67GdsW7715MEfRuP4pSZxJRJh6kj6Y48WRqVv4N1tRk/refs/rad/id -> refs/namespaces/z6Mkt67GdsW7715MEfRuP4pSZxJRJh6kj6Y48WRqVv4N1tRk/refs/rad/id
z6MknSLrJoTcukLrE435hVNQT4JUhbvWLX4kUzqkEStBU8Vi      worker: Git:  * [new ref]         refs/namespaces/z6Mkt67GdsW7715MEfRuP4pSZxJRJh6kj6Y48WRqVv4N1tRk/refs/rad/sigrefs -> refs/namespaces/z6Mkt67GdsW7715MEfRuP4pSZxJRJh6kj6Y48WRqVv4N1tRk/refs/rad/sigrefs
z6MknSLrJoTcukLrE435hVNQT4JUhbvWLX4kUzqkEStBU8Vi      worker: Fetch for rad:z42hL2jL4XNk6K8oHQaSWfMgCL7ji exited successfully
```

And, if we removed the line of `track_node(seed.id, Some("seed"))`,  the fetch will be skipped and test fails.

Reply:
Btw, I don't think this question is blocking this PR itself.

Reply:
Yeah, as I said, the seed is making its *own* announcement of local refs, to Bob, after it fetches from Alice. This is not Alice's announcement being relayed by the seed. That's why Bob is fetching from the seed, since he is tracking it.

But this isn't what we want actually, because we don't want Bob to have to track the seed. We want the seed to relay Alice's announcement to Bob, and Bob fetch from the seed because he tracks *Alice*.

The only way to achieve that, I think, is to actually not look at the announcer of the message, but rather at the refs themselves.

If the refs contain a remote that we are tracking, then we fetch from the announcer (the seed). Otherwise, we don't.

This should preserve most of the existing behavior, and allow us to revert some of the changes to the tests.

Reply:
A `RefsAnnouncement` contains multiple refs which are not necessarily from the same remote. If we check the tracking info of the remote instead of the announcer, we are changing the granularity of the check.

Check tracking of the announcer vs. check tracking of each ref is surely different behavior. One note:

Currently `RefsAnnouncement` implementation is different from [RIP 1](https://github.com/radicle-dev/rips/blob/master/0001-heartwood.md#reference-announcements): 

in code: 

```
pub struct RefsAnnouncement {
    /// Repository identifier.
    pub rid: Id,
    /// Updated refs.
    pub refs: BoundedVec<(NodeId, SignedRefs<Unverified>), REF_REMOTE_LIMIT>,
    /// Time of announcement.
    pub timestamp: Timestamp,
}
```

in RIP 1: 

```
RefsAnnouncement = (
    NodeId,
    RepositoryId,
    Map<RefName, CommitId>,
    Signature,
)
```

i.e. in RIP 1, there is an extra field `NodeId`.  It seems that it was extracted out, and becomes `announcer` in the code. Doesn't that mean the announcer should be the original node of the refs, not just a relayer or sender?

Reply:
Yes, we recently changed `RefsAnnouncement` (and I should update the RIP) to add more flexibility. Originally, you were only allowed to send announcements that you yourself signed, hence there was only one NodeId, since all refs were signed by that node.

Now, ref announcement messages can contain multiple remotes, each signed by the respective node. Hence the vector of (NodeId, SignedRefs) tuples.

Reply:
The difference is that this allows us to distinguish between the announcer of the refs, and the signer of the refs.

Reply:
> If we check the tracking info of the remote instead of the announcer, we are changing the granularity of the check.

Correct.

In the ideal case, we also only fetch the remotes that we're tracking. Eg. Seed announces refs from 5 peers, an I am only tracking 2 of those peers; I fetch only the 2 tracked peers from the announcer (if I'm connected to it).

Reply:
I have a related question:  in the following comments in the current code: 

https://github.com/radicle-dev/heartwood/blob/9d244b221a167c1c2bacbd00ecefc1a44ecab552/radicle-node/src/service.rs#L856-L858

What does `refs it has` mean?  Does it mean only the refs under the namespace of the announcer's public key, or all refs in the announcer's local storage (for that project)?

Reply:
> In the ideal case, we also only fetch the remotes that we're tracking. Eg. Seed announces refs from 5 peers, an I am only tracking 2 of those peers; I fetch only the 2 tracked peers from the announcer (if I'm connected to it).

Currently we either do fetch `All` or only one particular namespace. 
```
pub enum Namespaces {
    /// All namespaces.
    #[default]
    All,
    /// A single namespace, by public key.
    One(PublicKey),
}
```

To pick and choose the exact set of refs in an announcement, we will need to fetch refs of one peer at a time.

Reply:
I've made changes to the `Fetch` message. Now we will track the peer owning the refs, instead of the announcer. 

Hence this line in this test case is changed to track `bob` instead of `seed`.

Reply:
> To pick and choose the exact set of refs in an announcement, we will need to fetch refs of one peer at a time. And looks like we might need to change the Fetch message format on the wire as well to include namespaces.

That's actually what I'm working on here:
https://github.com/radicle-dev/heartwood/pull/354

Reply:
> What does refs it has mean? Does it mean only the refs under the namespace of the announcer's public key, or all refs in the announcer's local storage (for that project)?

All refs. Put differently, `/<alice>/ref/heads/master` and `/<bob>/refs/heads/master` are distinct refs. So in that comment an exact string match of the full ref string is what is meant.

Reply:
I've added the logic of checking the repo tracking scope and this test no longer needs code changes.

Reply:
> That's actually what I'm working on here:

Oops. I didn't know that. I have no problem if you want to land first.